domains/client/domain-executor/src/tests.rs

```diff
@@ -95,6 +95,40 @@ async fn collected_receipts_should_be_on_the_same_branch_with_current_best_block
 
     let parent_hash = *best_header.parent_hash();
 
+    let load_receipt = |primary_hash| -> Option<ExecutionReceipt<BlockNumber, Hash, Hash>> {
+        crate::aux_schema::load_execution_receipt::<_, _, BlockNumber, _>(
+            &*alice.backend,
+            primary_hash,
+        )
+        .unwrap_or_else(|err| panic!("Error occurred when loading execution receipt: {err}"))
+    };
+
+    // There is a chance that the receipt for #3 is not available at this point because
+    // writing the receipt happens in the background concurrently after the domain block #3
+    // notification is out.
+    //
+    // NOTE: There is a distinction between the domain block and regular block pipeline.
+    // - Regular block: fully imported to the blockchain database (block notification is fired at
+    //                  this point).
+    // - Domain block: fully imported to the blockchain database and the domain block
+    //                 post-hook (writing the receipt, etc) is finished.
+    //
+    // This difference should be okay in practice, we'll see if we need to take some actions in
+    // future and only take it into account in the tests for now.
+    let mut retries = 0;
+    loop {
+        if retries >= 10 || load_receipt(best_primary_hash).is_some() {
+            break;
+        }
+        tokio::time::sleep(std::time::Duration::from_millis(5)).await;
+        retries += 1;
+    }
+
+    assert!(
+        load_receipt(best_primary_hash).is_some(),
+        "Executor failed to process and generate the receipt of last primary block in 50ms"
```

Comment:
This may make the test unstable (depending on whether the CI is slow or not), how about calling `primary_node .notify_new_slot_and_wait_for_bundle` here. When the slot handling is finished, the previous block import task must be finished.

Reply:
I don't see how `primary_node.notify_new_slot_and_wait_for_bundle` would help, the bundle production and block processing run concurrently. As I have noted in the comments above, the domain block import notification does not mean the completion of a domain block pipeline. The potential here is when producing a bundle on new slot, the receipt is likely not yet available and then it fails at this slot https://github.com/subspace/subspace/actions/runs/4648954717/jobs/8226975359#step:7:963, what we want to is to always succeed in producing this bundle.

Reply:
Sorry, I mixed them up, but originally this situation should not happen.

Because, in the block import pipeline,  the `blocks_imported` notification actually comes before `blocks_importing` because `inner.import_block` (namely the `client.import_block` which sends the `every_import_notification` notification) is called before `acknowledgement_sender` is sent. 

And as `ferdie.produce_blocks(1)` will wait for the acknowledgment before completing, the `blocks_imported` notification must be handled since it is sent before `acknowledgement_sender` and we have set `primary_block_import_throttling_buffer_size` to 0 in the test.

But there are two places we did not consider:
- `tokio::select` randomly picks a branch to check first (see https://docs.rs/tokio/latest/tokio/macro.select.html#fairness), which means when both `every_import_notification` and `acknowledgement_sender` are sent, it will randomly pick one to handle instead of based on the order they comes.
- `futures::channel::mpsc::channel` actually have `buffer + num-senders` number of slot (see https://docs.rs/futures/latest/futures/channel/mpsc/fn.channel.html), which means even we have set `primary_block_import_throttling_buffer_size` to 0, it still has 1 slot thus acknowledgement may return before `blocks_imported` notification is handled

To fix that we may need to:
- Add `biased;` to the beginning of the `tokio::select` usage as the document suggested to ensure the order of branches to process, this should not affect the production code as we didn't have any assumption on this.
- Send one more `acknowledgement_sender` and wait for one more acknowledgement in the mock primary node, when this acknowledgement returns we can ensure `blocks_imported` notification must be handled.

WDYT @liuchengxu

Reply:
This sounds interesting, the reason now we have the term `ImportedBlock` and `BlockImporting` is in that we did run into #1143 which looked like the block is not fully imported at which point the notification from sc-consensus-subspace is fired, it's contrary to the analysis here (it makes some sense to me actually) that the notification of `ImportedBlock` is fired before `BlockImporting` in theory, if it does work as this way, we shouldn't encounter #1143. I suspect something from substrate or elsewhere I can't tell does not work as expected.

- Adding `biased` is okay to me, we won't set the buffer size to zero except for the tests anyway.
- I'm not sure about sending another ack. Having the buffer size as 0 and one sender means the channel will be blocked unless each received message is processed in my understanding. If the `ImportedBlock` notification always comes before `BlockImporting` with `biased`, then the corresponding `BlockImporing` notification won't be received until the notified primary block is fully processed by the executor, isn't this enough?

How about leaving this workaround as is so that this PR can move forward as it should be practically helpful to some extent IMO? And file a new dedicated as I believe there are multiple places that seemingly are affected, e.g., https://github.com/subspace/subspace/actions/runs/4657062650/jobs/8241286079?pr=1365#step:7:986,

Reply:
> if it does work as this way, we shouldn't encounter https://github.com/subspace/subspace/issues/1143. I suspect something from substrate or elsewhere I can't tell does not work as expected

Yeah, I feel the same... but I have added `biased` and some logs and then run a test multiple times, which always shows `ImportedBlock` is fired before `BlockImporting` thus I think we are good to proceed this step with the test.

> then the corresponding BlockImporing notification won't be received until the notified primary block is fully processed by the executor

It may not, `futures::channel::mpsc::channel(0)` can send 1 message without blocking, thus after `ImportedBlock ` is **received**, `BlockImporing` can be sent without blocking (so the `acknowledgement_sender` will also respond), and at the time `ImportedBlock` may not be fully processed.

> How about leaving this workaround as is so that this PR can move forward as it should be practically helpful to some extent IMO? And file a new dedicated as I believe there are multiple places that seemingly are affected

Sure! I will open a dedicated PR for the abovementioned changes.